File: libflatsurf/src/equivalence_class.cc
/**********************************************************************
 *  This file is part of flatsurf.
 *
 *        Copyright (C) 2022 Julian Rüth
 *
 *  Flatsurf is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  Flatsurf is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with flatsurf. If not, see <https://www.gnu.org/licenses/>.
 *********************************************************************/

#include <ostream>

#include "../flatsurf/equivalence_class.hpp"
#include "impl/equivalence_class_code.hpp"
#include "impl/equivalence.impl.hpp"

#include "impl/equivalence_class.impl.hpp"

namespace flatsurf {

template <typename Surface>
EquivalenceClass<Surface>::EquivalenceClass(const Surface& surface, const Equivalence<Surface>& equivalence) : self(spimpl::make_impl<ImplementationOf<EquivalenceClass>>(surface, equivalence)) {}

template <typename Surface>
bool EquivalenceClass<Surface>::operator==(const EquivalenceClass<Surface>& other) const {
  return *self->code == *other.self->code && self->equivalence == other.self->equivalence;
}

template <typename Surface>
size_t EquivalenceClass<Surface>::automorphisms() const {
  throw std::logic_error("not implemented: automorphisms()");
}

template <typename Surface>
const Surface& EquivalenceClass<Surface>::representative() const {
  // TODO: Canonicalize
  return self->surface;
}

template <typename Surface>
ImplementationOf<EquivalenceClass<Surface>>::ImplementationOf(const Surface& surface, const Equivalence<Surface>& equivalence) :
  surface(surface),
  equivalence(equivalence),
  code(equivalence.self->code(surface)) {}

template <typename Surface>
std::ostream& operator<<(std::ostream& os, const EquivalenceClass<Surface>& clazz) {
  return os << "[" << clazz.representative() << "] identified by (" << *clazz.self->code << ")";
}

}

// Instantiations of templates so implementations are generated for the linker
#include "util/instantiate.ipp"

LIBFLATSURF_INSTANTIATE_MANY_WRAPPED((LIBFLATSURF_INSTANTIATE_WITH_IMPLEMENTATION), EquivalenceClass, LIBFLATSURF_FLAT_TRIANGULATION_TYPES)

File: pyflatsurf/test/equivalence_class.py
#!/usr/bin/env python

######################################################################
# This file is part of flatsurf.
#
#       Copyright (C) 2022-2023 Julian Rüth
#
# flatsurf is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# flatsurf is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with flatsurf. If not, see <https://www.gnu.org/licenses/>.
######################################################################

import sys
import pytest

from pyflatsurf import flatsurf
import surfaces


def test_equivalence_class():
    for coefficients in ['long long', 'mpz_class', 'mpq_class', 'eantic::renf_elem_class', 'exactreal::Element<exactreal::IntegerRing>', 'exactreal::Element<exactreal::RationalField>', 'exactreal::Element<exactreal::NumberField>']:
        square = surfaces.square(flatsurf.Vector[coefficients])
        l = surfaces.L(flatsurf.Vector[coefficients])
        S = type(square)
        for equivalence in [flatsurf.Equivalence[S].combinatorial(), flatsurf.Equivalence[S].linear()]:
            a = flatsurf.EquivalenceClass[S](square, equivalence)
            b = flatsurf.EquivalenceClass[S](l, equivalence)

            assert a != b
            assert hash(a) != hash(b)
            assert a == flatsurf.EquivalenceClass[S](square, equivalence)
            assert hash(a) == hash(flatsurf.EquivalenceClass[S](square, equivalence))

if __name__ == '__main__': sys.exit(pytest.main(sys.argv))
